File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.py
from collections import defaultdict

class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        left, right = 0, 0
        max_len = 0
        majority = 0
        counts = defaultdict(int)

        while right < len(s):
            counts[s[right]] += 1
            majority = max(majority, counts[s[right]])
            while (right - left + 1) - majority > k:
                counts[s[left]] -= 1
                left += 1
            max_len = max(max_len, right - left + 1)
            right += 1

        return max_len
        

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int[] count = new int[26];      // freq in window (assumes 'A'..'Z')
        int left = 0, maxCount = 0, maxLen = 0;

        for (int right = 0; right < s.length(); right++) {
            int idx = s.charAt(right) - 'A';
            count[idx]++;
            maxCount = Math.max(maxCount, count[idx]);   // majority count in window

            // if replacements needed > k, shrink from left
            while ((right - left + 1) - maxCount > k) {
                count[s.charAt(left) - 'A']--;
                left++;
            }
            maxLen = Math.max(maxLen, right - left + 1);
        }
        return maxLen;
    }
}
